memstore/upsert_batch.go

```diff
@@ -520,97 +573,63 @@ func readUpsertBatchNew(buffer []byte) (*UpsertBatch, error) {
 
 }
 
-// TODO: delete after upsert batch new version migration
-func readUpsertBatchOld(buffer []byte) (*UpsertBatch, error) {
-	batch := &UpsertBatch{
-		buffer:      buffer,
-		columnsByID: make(map[int]int),
-	}
-
-	// numRows.
-	reader := utils.NewBufferReader(buffer)
-
-	numRows, err := reader.ReadInt32(0)
-	if err != nil {
-		return nil, utils.StackError(err, "Failed to read number of rows")
-	}
-	if numRows < 0 {
-		return nil, utils.StackError(err, "Number of rows %d should be >= 0", numRows)
-	}
-
-	batch.NumRows = int(numRows)
-	// numColumns.
-	numColumns, err := reader.ReadUint16(4)
-	if err != nil {
-		return nil, utils.StackError(err, "Failed to read number of columns")
-	}
-	batch.NumColumns = int(numColumns)
-
-	// Header too small, error out.
-	if len(buffer) < 8+common.ColumnHeaderSize(batch.NumColumns) {
-		return nil, utils.StackError(nil, "Invalid upsert batch data with incomplete header section")
-	}
-
-	header := common.NewUpsertBatchHeader(buffer[8:], batch.NumColumns)
-
-	columns := make([]*columnReader, batch.NumColumns)
-	for i := range columns {
-		columnType, err := header.ReadColumnType(i)
-		if err != nil {
-			return nil, utils.StackError(err, "Failed to read type for column %d", i)
-		}
+// ResolveEnumDict resolves upsert batch level enum dict with instance level enum dictionary
+func (u *UpsertBatch) ResolveEnumDict(table string, tableSchema *TableSchema, metaStore metastore.MetaStore) error {
+	tableSchema.RLock()
```

Comment:
better and utility function like GetEnumDict() and do the lock/unlock operation should be inside the object.

Reply:
we agreed early on the convention that the user will do lock and unlock consistently.